### ISSUE TITLE
Deadlock in IOApp when calling System.exit

### ISSUE THREAD

**Issue by scf37** [typelevel/cats-effect#746]
```
object ApiStarter extends IOApp {
  def run(args: List[String]): IO[ExitCode] = Resource.make(IO(System.exit(0)))(_ => IO(())).use { _ =>
    ???
  }
}
```

Hangs forever, clean exit expected.
It seems that it deadlocks on finalizer in IOAppPlatform.

cats-effect 2.0.0
cats 2.0.0

**Comment by scf37** [comment]
This one can only be reproduced when running from IDEA.
jstack output:

```
2020-01-10 18:12:40
Full thread dump Java HotSpot(TM) 64-Bit Server VM (25.181-b13 mixed mode):

"shutdownHook1" #12 prio=5 os_prio=0 tid=0x00007f2ad4004000 nid=0x3824 waiting on condition [0x00007f2adbefd000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x000000071e18f038> (a cats.effect.internals.IOPlatform$OneShotLatch)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
	at cats.effect.internals.IOPlatform$.$anonfun$unsafeResync$2(IOPlatform.scala:51)
	at cats.effect.internals.IOPlatform$$$Lambda$54/912672706.apply$mcV$sp(Unknown Source)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:57)
	at scala.concurrent.package$.blocking(package.scala:146)
	at cats.effect.internals.IOPlatform$.unsafeResync(IOPlatform.scala:51)
	at cats.effect.IO.unsafeRunTimed(IO.scala:324)
	at cats.effect.IO.unsafeRunSync(IO.scala:239)
	at cats.effect.internals.IOAppPlatform$.$anonfun$installHook$2(IOAppPlatform.scala:61)
	at cats.effect.internals.IOAppPlatform$$$Lambda$50/931496835.apply$mcV$sp(Unknown Source)
	at scala.sys.ShutdownHookThread$$anon$1.run(ShutdownHookThread.scala:37)

"ioapp-compute-0" #11 daemon prio=5 os_prio=0 tid=0x00007f2b42ac4000 nid=0x3823 in Object.wait() [0x00007f2adbffd000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x000000071dd80b10> (a scala.sys.ShutdownHookThread$$anon$1)
	at java.lang.Thread.join(Thread.java:1252)
	- locked <0x000000071dd80b10> (a scala.sys.ShutdownHookThread$$anon$1)
	at java.lang.Thread.join(Thread.java:1326)
	at java.lang.ApplicationShutdownHooks.runHooks(ApplicationShutdownHooks.java:107)
	at java.lang.ApplicationShutdownHooks$1.run(ApplicationShutdownHooks.java:46)
	at java.lang.Shutdown.runHooks(Shutdown.java:123)
	at java.lang.Shutdown.sequence(Shutdown.java:167)
	at java.lang.Shutdown.exit(Shutdown.java:212)
	- locked <0x000000071dd82630> (a java.lang.Class for java.lang.Shutdown)
	at java.lang.Runtime.exit(Runtime.java:109)
	at java.lang.System.exit(System.java:971)
	at mpp.ApiStarter$.$anonfun$run$1(ApiStarter.scala:30)
	at mpp.ApiStarter$$$Lambda$11/1361289747.apply$mcV$sp(Unknown Source)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:87)
	at cats.effect.internals.IORunLoop$.start(IORunLoop.scala:34)
	at cats.effect.internals.IOBracket$.$anonfun$apply$1(IOBracket.scala:44)
	at cats.effect.internals.IOBracket$.$anonfun$apply$1$adapted(IOBracket.scala:34)
	at cats.effect.internals.IOBracket$$$Lambda$27/1073763441.apply(Unknown Source)
	at cats.effect.internals.IORunLoop$RestartCallback.start(IORunLoop.scala:341)
	at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:119)
	at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:355)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:376)
	at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:316)
	at cats.effect.internals.IOShift$Tick.run(IOShift.scala:36)
	at cats.effect.internals.PoolUtils$$anon$2$$anon$3.run(PoolUtils.scala:51)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"Attach Listener" #10 daemon prio=9 os_prio=0 tid=0x00007f2af8001000 nid=0x3822 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Service Thread" #9 daemon prio=9 os_prio=0 tid=0x00007f2b40e4f000 nid=0x3815 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C1 CompilerThread2" #8 daemon prio=9 os_prio=0 tid=0x00007f2b40e38000 nid=0x3814 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread1" #7 daemon prio=9 os_prio=0 tid=0x00007f2b40e34800 nid=0x3813 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread0" #6 daemon prio=9 os_prio=0 tid=0x00007f2b40e32800 nid=0x3812 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Monitor Ctrl-Break" #5 daemon prio=5 os_prio=0 tid=0x00007f2b40e21800 nid=0x3810 runnable [0x00007f2b27a29000]
   java.lang.Thread.State: RUNNABLE
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:171)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	- locked <0x00000007198e9d80> (a java.io.InputStreamReader)
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	at java.io.BufferedReader.fill(BufferedReader.java:161)
	at java.io.BufferedReader.readLine(BufferedReader.java:324)
	- locked <0x00000007198e9d80> (a java.io.InputStreamReader)
	at java.io.BufferedReader.readLine(BufferedReader.java:389)
	at com.intellij.rt.execution.application.AppMainV2$1.run(AppMainV2.java:64)

"Signal Dispatcher" #4 daemon prio=9 os_prio=0 tid=0x00007f2b40192000 nid=0x380d runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Finalizer" #3 daemon prio=8 os_prio=0 tid=0x00007f2b4015e000 nid=0x3806 in Object.wait() [0x00007f2b2cb54000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x0000000719108ed0> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:144)
	- locked <0x0000000719108ed0> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:165)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:216)

"Reference Handler" #2 daemon prio=10 os_prio=0 tid=0x00007f2b4015b800 nid=0x3805 in Object.wait() [0x00007f2b2cc55000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x0000000719106bf8> (a java.lang.ref.Reference$Lock)
	at java.lang.Object.wait(Object.java:502)
	at java.lang.ref.Reference.tryHandlePending(Reference.java:191)
	- locked <0x0000000719106bf8> (a java.lang.ref.Reference$Lock)
	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:153)

"main" #1 prio=5 os_prio=0 tid=0x00007f2b40011800 nid=0x37fc waiting on condition [0x00007f2b47bd6000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x000000071dbe2f00> (a cats.effect.internals.IOPlatform$OneShotLatch)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
	at cats.effect.internals.IOPlatform$.$anonfun$unsafeResync$2(IOPlatform.scala:51)
	at cats.effect.internals.IOPlatform$$$Lambda$54/912672706.apply$mcV$sp(Unknown Source)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:57)
	at scala.concurrent.package$.blocking(package.scala:146)
	at cats.effect.internals.IOPlatform$.unsafeResync(IOPlatform.scala:51)
	at cats.effect.IO.unsafeRunTimed(IO.scala:324)
	at cats.effect.IO.unsafeRunSync(IO.scala:239)
	at cats.effect.internals.IOAppPlatform$.main(IOAppPlatform.scala:24)
	at cats.effect.IOApp.main(IOApp.scala:67)
	at cats.effect.IOApp.main$(IOApp.scala:66)
	at mpp.ApiStarter$.main(ApiStarter.scala:26)
	at mpp.ApiStarter.main(ApiStarter.scala)

"VM Thread" os_prio=0 tid=0x00007f2b40152000 nid=0x3804 runnable 

"GC task thread#0 (ParallelGC)" os_prio=0 tid=0x00007f2b40032800 nid=0x37fd runnable 

"GC task thread#1 (ParallelGC)" os_prio=0 tid=0x00007f2b40034000 nid=0x37fe runnable 

"GC task thread#2 (ParallelGC)" os_prio=0 tid=0x00007f2b40036000 nid=0x3801 runnable 

"GC task thread#3 (ParallelGC)" os_prio=0 tid=0x00007f2b40037800 nid=0x3802 runnable 

"VM Periodic Task Thread" os_prio=0 tid=0x00007f2b40e52800 nid=0x3816 waiting on condition 

JNI global references: 443
```

**Comment by gurinderu** [comment]
You forgot to mention it works without idea. In my opinion, this is not a cats bug.

**Comment by djspiewak** [comment]
My guess is that IntelliJ is doing something slightly odd with `System.exit`. I'd like to fix this, for sure, but it's hard to say without knowing what their semantics are with respect to running processes. Is it actually forked, or are they trying to sandbox it within the main process?

**Comment by scf37** [comment]
Forked.
```
Category	Value
VM Arguments	-javaagent:/home/asm/dev/idea-IU-192.7142.36/lib/idea_rt.jar=39535:/home/asm/dev/idea-IU-192.7142.36/bin -Dfile.encoding=UTF-8
```

Idea uses java agent.

Moreover, it reproduces only when using `run`, not when using `debug`. I suspect complicated concurrency bug, that's why I created this ticket.

**Comment by djspiewak** [comment]
Does it use the same agent when debugging? More relevantly, does that agent *do* the same things? Given all the madness that agents can perform, it's still really really difficult to say what may or may not be going on here. I suspect though that it's not really a concurrency bug, but rather just the way that IntelliJ is interfering with the functionality of shutdown hooks and maybe even `System.exit` itself.

**Comment by scf37** [comment]
Managed to reproduce it in debug mode as well. It seems that `fiber.cancel` in IOAppPlatform finalizer never ends, bug is reproduces only if state is not empty:
IOConnection.scala:

```
    val cancel = IO.suspend {
      state.getAndSet(null) match {
        case null | Nil =>
          IO.unit
        case list =>
          CancelUtils.cancelAll(list.iterator) // <-- breakpoint here
      }
    }
```

On some runs state is empty, on some runs it is not.

**Comment by scf37** [comment]
Successfully reproduced it when running assembled (using sbt-pack) application.
It seems that `sbt run` makes more disturbance to the app than idea!

Here is there repo with sample application and instructions: https://github.com/scf37/iohalt

**Comment by djspiewak** [comment]
Thanks for tracking down this reproduction! I agree this does seem like a problem in `IOApp`, likely in the way that shutdown handlers are being treated.

**Comment by scf37** [comment]
Some context to clarify importance of this bug:

When OOM occurs during resource initialization, `IOApp`'s `ExecutionContext` calls `System.exit()` which leads to deadlock. Currently we fix that but using `Runtime.getRuntime.halt()` but that bypasses other shutdown handlers, e.g. async logging flush.

**Comment by milanvdm** [comment]
We have the same issue with cats-effect 3. Our use case is that we want to restart our application in certain cases.

**Comment by vasilmkd** [comment]
@milanvdm Do you have a reproduction by any chance? I tried the one at the top of this issue, and it seems to work on `3.2.9`.

**Comment by vasilmkd** [comment]
Ah I see, running with `fork := true` seems to show the issue.

**Comment by vasilmkd** [comment]
The problem specifically is calling `System.exit(0)` from an uncancelable `IO` (in this issue specifically, this is bracket acquire).

**Comment by milanvdm** [comment]
Im glad you found a reproduction and sad that I couldnt give one to you! Thanks! :) If you need any help, let me know!

**Comment by vasilmkd** [comment]
@milanvdm You can work with me to set the expectations for this change. The issue is, as I said, observable when the main fiber of your application is generally uncancelable. Here's what happens.

Before the `IOApp` starts, Cats Effect registers a shutdown hook with the JVM. Any time Ctrl+C is sent, or `System.exit` is called, this shutdown hook gets triggered, which in turn cancels the main fiber of the `IOApp`. However, if the main fiber is uncancelable, the cancelation is backpressured (normal cancelation semantics of Cats Effect 3), and this will obviously go on forever, because no one is trying to make progress. If you try to run `System.exit` from a fiber which is uncancelable, it will work.

Afaik, it is common courtesy to try to let any process try and finish up within 10 seconds after the shutdown signal has been sent. Would a mechanism like that be acceptable in this case? The problem obviously comes from the fact that `System.exit` is a JVM method and can be called anywhere, so we cannot really add any sort of filter for it and have bespoke logic for it.

I would appreciate your thoughts.

**Comment by milanvdm** [comment]
This is a tricky one. Because purely based on expectations, a user would expect `IO(System.exit)` to stop the application from running.

One question: is there a specific reason the main fiber is uncancelable? Would it make sense to make it cancelable?

What about adding `IO.exit[ExitCase]`? This could be a specific case in the `runLoop` which triggers a proper shutdown of the main fiber. Although, not sure if it is even possible, does not solve the unexpected behavior of `IO(System.exit(1))`.

**Comment by vasilmkd** [comment]
The main fiber runs the `IO` that you give to `def run(args: List[String]): IO[ExitCase]`, and there's no point in putting restrictions as to whether it can be cancelable or uncancelable. For resource safety, you have to await the cancelation of any fiber, otherwise, you're leaking unclosed resources.

`System.exit` is magic, and calling it manually circumvents the `ExitCode`, because the app is literally not allowed to reach it and exit.

If you want to exit an app "early", just return an `ExitCode` at some condition, instead of calling `System.exit`.

**Comment by milanvdm** [comment]
In our case, the fiber is running in the background. So the restart is triggered from a fiber running in the background of the main application. So I don't see a way to exit the app early that way with just an `ExitCode`.

**Comment by vasilmkd** [comment]
Would something like this work? `fs2` has even better features like `SignallingRef` which can interrupt streams.

```scala
object JVMExit extends IOApp {
  def run(args: List[String]): IO[ExitCode] =
    for {
      deferred <- Deferred[IO, ExitCode]
      _ <- (IO.sleep(10.seconds) *> deferred.complete(ExitCode.Success)).start
      foreground <- (IO.sleep(1.second) *> IO.println("working")).foreverM.start
      exit <- deferred.get
      _ <- foreground.cancel
    } yield exit
}
```

**Comment by vasilmkd** [comment]
@milanvdm We will still iterate on this issue, I will probably add configuration for how long to await cancelation on shutdown (if at all) and this will still help you without having to modify your code (apart from the config). I will prepare a PR later tonight.
